Refactor CurlHandler for threading and resource cleanup

stephentoub opened the pull request:
@davidsh, @CIPop, @pgavlin: The first commit in this PR enables the System.Net.Requests tests to run on Unix in CI.  Just a few minor tweaks.

@kapilash: The second commit is your commit with new tests from #2916, which this PR replaces.

@kapilash, @vijaykota: The third commit is the majority of the changes for this PR...

CurlHandler is now split into three types:
- CurlHandler, which represents the shell of the handler, contains some helper functions, and contains a "MultiAgent".  It submits "EasyRequest"s to the "MultiAgent".
- CurlHandler.EasyRequest, which represents all of the state associated with a SendAsync request on the HttpClient.
- CurlHandler.MultiAgent, which represents the owner of a libcurl multi handle and all of the processing associated with it.  The MultiAgent is responsible for actually processing EasyRequests, which CurlHandler queues to it.

MultiAgent is designed to not require any cleanup.  When the first EasyRequest arrives, it spins up a worker that creates a multi handle to process the request.  Any additional requests to arrive while that worker is active will be processed by that worker and multi handle.  When no work remains, the multi handle and the worker goes away.  If/when another request arrives, a new multi handle / worker are spun up.  Etc.

Almost all of the locking previously there was removed, as the worker thread is now solely responsible for doing all work on the multi handle; any requests for processing come in through the requests queue processed by that thread. This design avoids threading issues we faced with regards to accessing state managed by the multi handle without holding the lock (e.g. polling file descriptors) and avoids having multiple threads blocked waiting on the lock to do processing associated with the multi handle.  It also makes resource cleanup much easier and more deterministic: all state associated with the easy handle is cleaned up when that thread is done processing the easy handle, and all state associated with the multi handle is cleaned up when the worker thread shuts down. The change also ensures that we always complete the response message, and that any exceptions that occur are propagated as needed to Read on the response stream (fixes #3015).

Further, the design will make it trivial in the future to change the relationship between a CurlHandler and a MultiAgent:
- If we want one MultiAgent shared across all CurlHandlers, we just change the _agent field to be static instead of instance.
- If we want a pool of MultiAgents, we can simply create a collection of them and round-robin between them from one or more CurlHandlers.
- Etc.

All of the configuration code remains, e.g. what options are set on the easy and multi handles, but the code has been refactored.  I also include CURLOPT_NOSIGNAL to avoid multithreading problems per the docs.

Along the way, I cleaned up some stale interop that's gotten left behind through various iterations, and I fixed a few small issues related to behavioral differences from Windows, e.g. CookieContainer never returning null, UseCookies defaulting to true, etc.  (Fixes #3016)

stephentoub commented:
Thanks, David.

vijaykota commented on src/Common/src/Interop/Unix/libcurl/Interop.libcurl_multi.cs:
```diff
@@ -46,16 +32,16 @@ internal static partial class libcurl
 
         [DllImport(Interop.Libraries.LibCurl)]
-        public static extern int curl_multi_socket_action(
+        public static extern int curl_multi_perform(
```
The reason for choosing curl_multi_socket_action over curl_multi_perform was the foll. from section labelled MULTI_SOCKET(http://curl.haxx.se/libcurl/c/libcurl-multi.html)

>curl_multi_socket_action function offers a way for applications to not only avoid being forced to use select(), but it also offers a much more high-performance API that will make a significant difference for applications using large numbers of simultaneous connections. 

Currently we have used a platform-agnostic poll with lower perf but the hope is that we can move to a platform-specific performant version (maybe epoll on Linux etc.)

stephentoub replied:
Yes, I have a comment on that here: https://github.com/dotnet/corefx/pull/3050/files#diff-e5ffb568f6f994c4243d35ef1b6830ceR282

vijaykota commented on src/Common/src/System/Net/Http/HttpHandlerDefaults.cs:
```diff
@@ -11,5 +11,7 @@ internal static class HttpHandlerDefaults
         public const int DefaultMaxAutomaticRedirections = 50;
         public const DecompressionMethods DefaultAutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
         public const bool DefaultAutomaticRedirection = true;
+        public const bool DefaultUseCookies = true;
```
@kapilash, was the earlier default of false incorrect?

kapilash replied:
Yes. It should be true. 
With the default value set to true, we may have another issue #3016 I haven't read the code completely

stephentoub replied:
@kapilash, if you're referring to the cookie container coming back as null rather than as a default instance, I addressed that in the CookieContainer property by lazily initializing the collection.

kapilash commented:
Thanks @stephentoub 

I believe the hang issue, #3015, will be resolved with this.

However, when I look at the [console output of the job](http://dotnet-ci.cloudapp.net/job/dotnet_corefx_linux_debug_prtest/3568/console), I think the tests haven't run.
Am I looking at the wrong output? I see that tests have been enabled for linux.

I believe  @davidsh haven't completely reviewed the tests. He was perhaps waiting for a consistent pass on #2916. If There are many comments on the tests, please feel free to ignore the commit, I will send out a separate one. Or please suggest what is the right approach :-)

Maxwe11 commented:
>Am I looking at the wrong output?

yes
see http://dotnet-ci.cloudapp.net/job/dotnet_corefx_linux_debug_tst_prtest/2203/console

kapilash commented:
Thanks @Maxwe11

vijaykota commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs:
```diff
@@ -0,0 +1,373 @@
+            [Conditional("CURLHANDLER_VERBOSE")]
+            private void SetDebugging()
+            {
+                SetCurlOption(CURLoption.CURLOPT_VERBOSE, 1L);
```
I think we can make this more useful by defining a CURLOPT_DEBUGFUNCTION

stephentoub replied:
Potentially.  My understanding is that if you enable verbose but not supply a custom debug function, you essentially get a debug function that writes to stderr the data associated with any CURLINFO_TEXT, CURLINFO_HEADER_OUT, and CURLINFO_HEADER_IN messages.  Those are basically all of the text-based cases.  There are then the CURLINFO_DATA_* and CURLINFO_SSL_DATA_* binary messages, which at the moment don't seem as useful when simply dumped to stderr.  If we find a need in the future, we can certainly add our own custom debug function to route all of this information elsewhere, to handle those additional messages, etc.

vijaykota replied:
Makes sense to me

vijaykota commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:
```diff
@@ -0,0 +1,658 @@
+                        // complexity would let us turn the O(N) operation into an O(1) operation.  The additional complexity would come
+                        // not only in the form of additional callbacks and managing the socket collection, but also in the form of timer
+                        // management, which is necessary when using the curl_multi_socket_* APIs and which we avoid by using just
+                        // curl_multi_wait/perform.
```
Ok didn't see this comment earlier. But isn't it better to code with curl_multi_socket* upfront?

stephentoub replied:
Until we do use epoll, there's little value in the extra complexity brought by switching to ```curl_multi_socket_*```... in fact there's extra cost to it because we end up needing to make many more P/Invoke calls and the like.  I think if we're going to do the switch, it makes sense to hold off on that until a separate PR that switches from ```curl_multi_perform/wait``` to ```curl_multi_socket_*``` and ```epoll``` all together, otherwise we're paying without any gains.  And such a switch should be a relatively simple drop-in replacement with this revised design:
- At the beginning of the worker after creating the multi handle, we epoll_create an epoll instance.
- We register for the socket callback function, and when it's invoked (synchronously, so no extra locking) we use epoll_ctl with EPOLL_CTL_ADD/DEL to add or remove the socket from being tracked.
- We register the timer functions as was done previously, but when the timer fires, it simply calls RequestWakeup to make sure the processing loop goes around again.
- Instead of curl_multi_wait, we use epoll.
- Instead of curl_multi_perform, we iterate through the set fds from epoll and use curl_multi_socket_action.  Etc.
- When the worker is about to end, we delete the epoll instance.
- Etc.

It's very possible we'll want to pull back in some of the deleted code (e.g. the timer interop stuff), but we can grab that from the history, and if/when we do want to do this, we can layer it on top of a foundation that I think puts us in a more reliable place.

Let me know if you disagree.  If you think it's really important to do as part of this PR, I can look at adding an additional commit to do it.

vijaykota replied:
I think the PR can go ahead and we will try to get data on whether curl_multi_socket_action+poll is better than curl_multi_perform+curl_multi_wait. Since only benefit of former seems to be that it works with a specific fd, we need to investigate if this is actually translating to O(n) vs 2*O(n) curl_multi_perform might have some inbuilt optimizations of its own

vijaykota commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:
```diff
@@ -0,0 +1,658 @@
+
+                            // Wait for a reader
+                            // TODO: The below call blocks till all the data has been read since
+                            //       response body is not suppored to be buffered in memory.
```
Nit: Typo "suppored"

stephentoub replied:
Thanks; this was just copied over from the original.  I'll fix it.

stephentoub commented:
> There are many comments on the tests, please feel free to ignore the commit, I will send out a separate one. Or please suggest what is the right approach :-)

@kapilash, I see, I thought you'd addressed all of the feedback.  Before submitting, I can remove that commit from this PR if you'd like, though I will need to keep the changes to the libcurl library name so that the System.Net.Requests tests run successfully.

vijaykota commented on src/Common/src/Interop/Unix/libcurl/Interop.libcurl_types.cs:
```diff
@@ -195,6 +166,20 @@ internal struct curl_version_info_data
+        };
+#pragma warning restore 0649
+
         public delegate int curl_socket_callback(
```
Nit: Probably not used anymore

stephentoub replied:
You're right, will delete.  We can bring it back along with some of the others if/when we switch to epoll.

vijaykota commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs:
```diff
@@ -1,505 +0,0 @@
-
-            lock (multiHandle)
-            {
-                multiHandle.Timer.Change((int)timeoutInMilliseconds, Timeout.Infinite);
```
You can modify project.json to remove reference to System.Threading.Timer

stephentoub replied:
Ok, will do.

vijaykota commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:
```diff
@@ -0,0 +1,658 @@
+                }
+
+                EasyRequest easy;
+                if (TryGetEasyRequestFromContext(context, out easy))
```
Good to check for cancellation (maybe in TryGetEasy itself)

stephentoub replied:
We could, but I'm not convinced it's necessary.  We're now registering for cancellation notifications via CancellationToken.Register.  If cancellation is requested, the curl_multi_wait will wake up, loop around, and as part of processing requests will see that cancellation was requested and will do all of the appropriate cleanup.  Is there a case I'm missing?  Is the concern that this will end up doing lots of work before we loop around again?

vijaykota replied:
I agree that it is not very important but seems like a cheap check.. my point was only for the tiny window between:
1. curl_multi_wait wakes up because data can be written
2. Cancellation is requested
3. CurlSendCallback is invoked

I guess we need to make the Stream.Read also cancellable for a proper fix covering other scenarios.

stephentoub replied:
> I guess we need to make the Stream.Read also cancellable for a proper fix covering other scenarios.

Yeah. My implementation of ReadAsync in https://github.com/dotnet/corefx/pull/3068 allows the read to be canceled via the CancellationToken.

vijaykota replied:
This Stream refers to the <b>request</b> content stream

stephentoub replied:
Oh, you were talking about the Stream in CurlSendCallback?  I missed that, as this whole comment sequence is on CurlReceiveHeadersCallback, not CurlSendCallback.

kapilash commented:
> @kapilash, I see, I thought you'd addressed all of the feedback. 

I did resolve all the comments that were made so far. There aren't any open issues other than that tests were hanging.  With this PR, (specifically, with curl_multi_perform) the tests will not hang.

stephentoub commented:
> I did resolve all the comments that were made so far.

Ok, good.  @davidsh, are you ok with the commit I cherry-picked from Hemanth's PR?  Any additional changes you'd like to see there?

vijaykota commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:
```diff
@@ -0,0 +1,658 @@
+                    // Transfer data from the request's content stream to libcurl
+                    try
+                    {
+                        byte[] arr = easy._associatedMultiAgent._transferBuffer;
```
Can the same buffer be used across all handles?

stephentoub replied:
Everything on the multi handle is processed by the single thread.  There shouldn't be any problem using a single buffer, and this saves us from having to allocate a 16KB buffer per request, instead doing it per agent and reusing that buffer across however many requests it processes.  Is there a particular case you're concerned about?

vijaykota replied:
CurlSendCallback will be invoked from a diff. thread not the thread that is running Process(). Or is behavior different with curl_multi_perform?

stephentoub replied:
> CurlSendCallback will be invoked from a diff. thread not the thread that is running Process

That's not the case.  It's always invoked from the thread calling Process, which within the call to curl_multi_perform.  This is confirmed by the FAQ (http://curl.haxx.se/docs/faq.html#Does_libcurl_use_threads):

"5.18 Does libcurl use threads?
Put simply: no, libcurl will execute in the same thread you call it in. All callbacks will be called in the same thread as the one you call libcurl in."

vijaykota replied:
Isn't this a problem if the read/write block? @kapilash, in your socket_action vs multi_perform test, can you check if the behavior is same with curl_multi_socket_action as well?

stephentoub replied:
Vijay, this was already a problem. No matter what design is employed, all work associated with the same multi handle must be serialized. If you block whatever thread that is, whether it's a dedicated thread or a thread holding a lock, the problem exists.

vijaykota commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:
```diff
@@ -0,0 +1,658 @@
+                                // set of descriptors.
+                                Interop.Sys.Close(thisRef._wakeupRequestedPipeFd);
+                                Interop.Sys.Close(thisRef._requestWakeupPipeFd);
+                                thisRef._wakeupRequestedPipeFd = 0;
```
Nit: Can move this right after the Close

stephentoub replied:
You mean move it up one line?  Ok, will do.

vijaykota commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:
```diff
@@ -0,0 +1,658 @@
+                    // curl_multi_wait, and we want to make sure it wakes up to see that there additional
+                    // requests waiting to be handled.  So we write to the wakeup pipe.
+                    Debug.Assert(_incomingRequests.Count >= 1, "We just queued a request, so the count should be at least 1");
+                    if (_incomingRequests.Count == 1)
```
Don't we need this for other transitions as well (eg. 1 to 2) so that curl_multi_wait updates its polling set

kapilash commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:
```diff
@@ -0,0 +1,658 @@
+                else // _workerRunning == true
+                {
+                    // The worker is already running.  If there are already queued requests, we're done.
+                    // However, if there aren't any queued requests, Process could be blocked inside of
```
Under what circumstances could a process be blocked in curl_multi_wait? Wouldn't it return under the FailSafeTimeOut?

stephentoub replied:
> Wouldn't it return under the FailSafeTimeOut?

Yes, but that timeout is long (and we potentially might want to make it longer).  That timeout is really there as a "failsafe", just in case something goes wrong, but really shouldn't be necessary.  This mechanism is meant to be the primary way that we wake up the waiting.  If we didn't have this, we'd need to make the timeout much shorter, in which case we'd have a thread polling and doing a busy waiting loop, e.g. waking up 30 times a second to see if there's anything to be done.  I currently think this strikes a nice compromise, but we can always revisit it later.

vijaykota replied:
Yes, it is difficult to choose a right value. Having this current longer value is a better choice

davidsh commented:
@stephentoub 

>Ok, good. @davidsh, are you ok with the commit I cherry-picked from Hemanth's PR? Any additional changes you'd like to see there?

I'm fine with the cherry-picks.

stephentoub commented:
Thanks, all.  I'm going to go ahead and merge this.  I'm happy to address any additional feedback in a subsequent PR.

vijaykota commented:
LGTM - I only have one point/doubt regarding the buffer used in CurlSendCallback

Thanks @stephentoub for the tip on CURL_WRITEFUNC_PAUSE.
